### PR TITLE
✨ config-gen create and bind to a non-default service account

### PR DESCRIPTION
**Pull request by pwittrock**
`kubebuilder alpha config-gen` support for non-default service accounts -- originally introduced in #2070

From #2070 description:

> As of now, all managers are deployed under the same service account in a namespace. Although this namespace is unique to the manager initially, it is a common admin pattern to deploy multiple managers under the same namespace for security reasons. This pattern becomes insecure given the current one-service-account-for-all approach, since RBAC will conflict.

> This PR adds a ServiceAccount (config/rbac/service_account.yaml) and changes the default service account name from default to controller-manager such that a user can specify which service account their manager should be created in. They may change this value by updating files referencing the metadata.name value in service_account.yaml.

> pkg/plugins/golang/v3: update all scaffold referencing the default service account with references to the "controller-manager" service account, specified in service_account.yaml.

### CODE REVIEW

File: pkg/cli/alpha/config-gen/cmd.go [hunk 1]
```diff
@@ -172,6 +172,13 @@ The KubebuilderConfigGen resource has the following fields:
 		# generate prometheus ServiceMonitor resource
 	    enableServiceMonitor: true
 
+      serviceAccount:
+	    # configure the service account used for RBAC
+	    name: foo
+
+		# generate the service account resource
+        generate: true
+
```
Comment (Adirio) [inline pkg/cli/alpha/config-gen/cmd.go:181]: Spaces and tabs

File: go.mod [hunk 2]
```diff
@@ -4,6 +4,7 @@ go 1.15
 
 require (
 	github.com/cloudflare/cfssl v1.5.0 // for `kubebuilder alpha config-gen`
+	github.com/go-logr/logr v0.3.0 // indirect
```
Comment (Adirio) [inline go.mod:7]: Try deleting these 3 imports and check if `go mod tidy` adds them.

File: pkg/cli/alpha/config-gen/templates/resources/service-account.template.yaml [hunk 3]
```diff
@@ -0,0 +1,7 @@
+{{ if not .Spec.ControllerManager.ServiceAccount.NoGenerate }}
+apiVersion: v1
+kind: ServiceAccount
+metadata:
+  name: {{ .Spec.ControllerManager.ServiceAccount.Name }}
+  namespace: {{ .Namespace }}
+{{ end }}
```
Comment (Adirio) [inline pkg/cli/alpha/config-gen/templates/resources/service-account.template.yaml:7]:
```suggestion
{{ end }}

```

File: pkg/cli/alpha/config-gen/testdata/serviceaccount/config.yaml [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+metadata:
+  name: simple
+spec:
+  crds:
+    sourceDirectory: ../project/...
+
+  controllerManager:
+    image: example/simple:latest
+    serviceAccount:
+      name: foo
```
Comment (Adirio) [inline pkg/cli/alpha/config-gen/testdata/serviceaccount/config.yaml:12]:
```suggestion
      name: foo

```

File: pkg/cli/alpha/config-gen/testdata/serviceaccountgenerate/config.yaml [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+  name: simple
+spec:
+  crds:
+    sourceDirectory: ../project/...
+
+  controllerManager:
+    image: example/simple:latest
+    serviceAccount:
+      name: foo
+      generate: true
```
Comment (Adirio) [inline pkg/cli/alpha/config-gen/testdata/serviceaccountgenerate/config.yaml:13]:
```suggestion
      generate: true

```

File: pkg/cli/alpha/config-gen/cmd.go (hunk 1 above)
Comment (pwittrock) [reply pkg/cli/alpha/config-gen/cmd.go:181]: Done

File: go.mod (hunk 2 above)
Comment (pwittrock) [reply go.mod:7]: Running the tests does, but building the binary doesn't.  Will delete.

**Comment by pwittrock** [comment]
PTAL

File: pkg/cli/alpha/config-gen/cmd.go [hunk 6]
```diff
@@ -148,54 +148,68 @@ The KubebuilderConfigGen resource has the following fields:
...
+        name: foo
+
+        # generate the service account resource
+        generate: true
+
+      serviceAccount:
+	    # configure the service account used for RBAC
+	    name: foo
+
+		# generate the service account resource
```
Comment (Adirio) [inline pkg/cli/alpha/config-gen/cmd.go:186]: I found at least these 4 tabs
```suggestion
      # configure the service account used for RBAC
      name: foo

    # generate the service account resource
```

**Review by Adirio — CHANGES_REQUESTED** [review]
IIRC, we are using a fixed name for the `ServiceAccount` and then prefixing it with a unique prefix per manager. This approach is a bit different, unsing a user-provided name and defaulting to the `KubebuilderConfigGen` name.

I don't have a strong opinion on which approach is better. @estroz?

File: pkg/cli/alpha/config-gen/cmd.go [hunk 7]
```diff
@@ -148,54 +148,68 @@ The KubebuilderConfigGen resource has the following fields:
...
+        # generate prometheus ServiceMonitor resource
+        enableServiceMonitor: true
+
+      serviceAccount:
+        # configure the service account used for RBAC
+        name: foo
+
+        # generate the service account resource
+        generate: true
+
```
Comment (estroz) [inline pkg/cli/alpha/config-gen/cmd.go:181]: Duplicate?

**Comment by estroz** [comment]
@Adirio correct. @pwittrock it seems some resources are generated [with their final name](https://github.com/kubernetes-sigs/kubebuilder/blob/master/pkg/cli/alpha/config-gen/templates/resources/rbac.template.yaml#L4), and some [do not have a name prefix](https://github.com/kubernetes-sigs/kubebuilder/blob/master/pkg/cli/alpha/config-gen/templates/resources/controller-manager.template.yaml#L4) such that they can be `kustomize edit set nameprefix`'d. Is this intentional?

**Comment by pwittrock** [comment]
> @Adirio correct. @pwittrock it seems some resources are generated with their final name, and some do not have a name prefix such that they can be kustomize edit set nameprefix'd

What is the issue that keeps nameprefix from working?  This could be an issue with the transformer plugin ordering -- does it work if you use the plugin as a base for the kustomization that does the prefix?

**Comment by pwittrock** [comment]
> IIRC, we are using a fixed name for the ServiceAccount and then prefixing it with a unique prefix per manager. This approach is a bit different, unsing a user-provided name and defaulting to the KubebuilderConfigGen name.

I suspect this is in part because of kustomize's inability to finely configure name transformations -- e.g. Kustomize doesn't allow a user to provide a name and then generate names from that, it only supports adding either a prefix or suffix.

**Comment by estroz** [comment]
>What is the issue that keeps nameprefix from working?

@pwittrock nothing necessarily, I'm just wondering why some templates are generated with their final name and some are not such that kustomize can prefix their name. Shouldn't they all be prefixed by kustomize, or all templates have the final name, ex. [`{{ .Metadata.Name }}-controller-manager`](https://github.com/kubernetes-sigs/kubebuilder/blob/master/pkg/cli/alpha/config-gen/templates/resources/controller-manager.template.yaml#L4)?

>This could be an issue with the transformer plugin ordering -- does it work if you use the plugin as a base for the kustomization that does the prefix?

It seems kustomize does not apply builtin functions to resources post transformation within the same kustomize dir. In #2099 I place the transformer in `config/configgen/kustomization.yaml` then reference that kustomize dir from `config/default/kustomization.yaml` as a resource; the latter contains a name prefix and applies it successfully.

**Pull request closed without merging**